### PR TITLE
Add tag filtering to challenge page

### LINKED ISSUE #22: Filter challenges by tag
Client side interface

### PR DESCRIPTION
**Pull request by zwade**
[Resolves #22] Colors are placeholder, however functionality is complete.

### CODE REVIEW

File: templates/base.html [hunk 1]
```diff
@@ -11,6 +11,8 @@
     <link href="/third_party/angular/angular-csp.css" rel="stylesheet">
     <link href="/css/scoreboard.css" rel="stylesheet">
     <link href="/css/scoreboard-colors.css" rel="stylesheet">
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons"
```
Comment (Matir) [inline templates/base.html]: Please use a protocol-relative URL

File: static/js/controllers/challenges.js [hunk 2]
```diff
@@ -106,16 +106,25 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
       $scope.currChall = null;
+      $scope.shownTags = {};
       var refresh = function() {
           categoryService.getList(function(data) {
               $scope.categories = data.categories;
           });
       };
 
+      tagService.getList(function(tags) {
+        $scope.allTags = tags.tags
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 3]
```diff
@@ -106,16 +106,25 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+        for (var i = 0; i < $scope.allTags.length; i++) {
+          $scope.shownTags[$scope.allTags[i].tagslug] = 2
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 4]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
           return scoreService.getCurrentPoints(chall) + " points";
       };
 
+      $scope.tagsAllowed = function(chall) {
+        for (var i = 0; i < chall.tags.length; i++) {
+          var type = $scope.shownTags[chall.tags[i].tagslug]
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 5]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+          if (type == 0) {
+            return false
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 6]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+          }
+        }
+
+        for (var i = 0; i < chall.tags.length; i++) {
+          var type = $scope.shownTags[chall.tags[i].tagslug]
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 7]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+          if (type == 2) {
+            return true
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 8]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+          }
+        }
+        return false;
+      }
+
+      $scope.toggleTag = function(t, click) {
+        var tindex = $scope.shownTags[t]
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 9]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+        //Return next permutation
+        if (click == 0) {
+          tindex += 1
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 10]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+        } else {
+          tindex += 3-1
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 11]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+        }
+        $scope.shownTags[t] = tindex % 3
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 12]
```diff
@@ -128,6 +137,39 @@ challengeCtrls.controller('ChallengeGridCtrl', [
...
+      }
+
+      $scope.getSentiment = function(tag) {
+        var sentiments = ['sentiment_dissatisfied', 'sentiment_neutral', 'sentiment_satisfied']
+        return sentiments[$scope.shownTags[tag.tagslug]]
```
Comment (Matir) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

File: static/js/controllers/challenges.js [hunk 13]
```diff
@@ -128,6 +137,47 @@ challengeCtrls.controller('ChallengeGridCtrl', [
           return scoreService.getCurrentPoints(chall) + " points";
       };
 
+      $scope.tagsAllowed = function(chall) {
+        var containsTag = function(chall, tagslug) {
+            for (var i = 0; i < chall.tags.length; i++) {
+              if (chall.tags[i].tagslug == tagslug) return true
```
Comment (zwade) [inline static/js/controllers/challenges.js]: Please end statements with a semicolon.

**Pull request merged**